server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java

```diff
@@ -304,12 +307,62 @@ private synchronized void updateSkipUnavailable(String clusterAlias, Boolean ski
 
-    public void updateRemoteClusterCredentials(Settings settings) {
-        remoteClusterCredentialsManager.updateClusterCredentials(settings);
+    public synchronized void updateRemoteClusterCredentials(Supplier<Settings> settingsSupplier, ActionListener<Void> listener) {
+        updateRemoteClusterCredentials(settingsSupplier.get(), listener);
     }
 
-    public RemoteClusterCredentialsManager getRemoteClusterCredentialsManager() {
-        return remoteClusterCredentialsManager;
+    public synchronized void updateRemoteClusterCredentials(Settings settings, ActionListener<Void> listener) {
+        final UpdateRemoteClusterCredentialsResult result = remoteClusterCredentialsManager.updateClusterCredentials(settings);
+        // We only need to rebuild connections when a credential was newly added or removed for a cluster alias, not if the credential
+        // value was updated. Therefore, only consider added or removed aliases
+        final int totalConnectionsToRebuild = result.addedClusterAliases().size() + result.removedClusterAliases().size();
+        if (totalConnectionsToRebuild == 0) {
+            logger.debug("no connection rebuilding required after credentials update");
+            listener.onResponse(null);
+            return;
+        }
+        logger.info("rebuilding [{}] connections after credentials update", totalConnectionsToRebuild);
+        final GroupedActionListener<Void> groupedListener = new GroupedActionListener<>(
+            totalConnectionsToRebuild,
+            listener.map(responses -> {
+                logger.info("rebuild complete for [{}] connections after credentials update", responses.size());
+                return null;
+            })
+        );
+        for (var clusterAlias : result.addedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+        for (var clusterAlias : result.removedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+    }
+
+    private void maybeRebuildConnectionOnCredentialsChange(
+        String clusterAlias,
+        Settings settings,
+        GroupedActionListener<Void> groupedListener
+    ) {
+        if (false == remoteClusters.containsKey(clusterAlias)) {
+            // A credential was added or removed before a remote connection was configured.
+            // Without an existing connection, there is nothing to rebuild.
+            logger.info("no connection rebuild required for remote cluster [{}] after credentials change", clusterAlias);
+            groupedListener.onResponse(null);
+            return;
+        }
+
+        try {
+            final PlainActionFuture<RemoteClusterConnectionStatus> future = new PlainActionFuture<>();
+            updateRemoteCluster(clusterAlias, settings, true, future);
+            final RemoteClusterConnectionStatus status = future.actionGet(10, TimeUnit.SECONDS);
```

Comment:
I do not like stacking the timeouts one after the other like this. If we have, say, 150 remotes, and they all hit the 10s timeout, then with this logic we'll be waiting for 1500s = 25 minutes. If we need to time out at all then we should have one top-level timeout, and TBH I'd rather it was in the caller seeing as how we don't do anything (except logging) when we hit the timeout anyway.

I know we have this problem when applying cluster states too, and IMO that's a bug. It's tricky to solve in that case because of the mechanics of how settings are applied, but I would rather avoid making the same mistake here too.

Reply:
Fair point. I don't feel strongly about the per-connection timeout (nor the top-level timeout for that matter). 

If we have 150 remotes, and each connection takes >10s to rebuild the overall request will still wait for 25m. A top level timeout could prevent this but it's difficult to pick a reasonable default since I expect timeouts to vary greatly based on the network situation for a given connection (@jakelandis points this out as [here](https://github.com/elastic/elasticsearch/pull/103460#discussion_r1434495042) which prompted me to consider a per-connection timeout) -- e.g., a healthy connection will rebuild in >1s but a slow/unstable network could result in long waits.

It could be nice to add a timeout parameter to the initial, generic `reload-secure-settings` request but we have no way of passing this down to the individual plugin reload call, since the [reload()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java#L134) call only takes settings as input and I would prefer not to change the `ReloadablePlugin` interface here since that's a bigger change that requires separate input and deliberation.  

The only other alternative that comes to mind is adding a dedicated setting where end-users can set a connection rebuild timeout but that feels hacky and like an overkill since, as you point out, we don't do anything (except logging) when we hit the timeout anyway.

Given that it's hard to come up with a generic top-level timeout, and that individual connection timeouts can give us very high (and pretty arbitrary) wait times, I prefer to remove the timeout altogether for now and possibly revisit this in the future. @DaveCTurner @jakelandis let me know if you have objections.

Reply:
I would be happy with no timeout here I think, especially if we trigger all the connection attempts in parallel since there's other timeouts in play lower down the stack, preventing this from really waiting forever.

I do think it's worth reconsidering the synchronous nature of the `reload()` API, which forces us to call each plugin in turn (or do something worse like spawning a thread per plugin). Stuff like this should be async all the way down the stack IMO and then these concerns wouldn't really arise.